pkg/catalog/manager/catalog_common.go

```diff
@@ -0,0 +1,123 @@
+package manager
+
+import (
+	"fmt"
+
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/client/management/v3"
+	"github.com/sirupsen/logrus"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func (m *Manager) deleteTemplates(key string, obj *v3.Catalog) error {
+	templates, err := m.getTemplateMap(obj.Name)
+	if err != nil {
+		return err
+	}
+	tvToDelete := map[string]struct{}{}
+	for _, t := range templates {
+		tvs, err := m.getTemplateVersion(t.Name)
+		if err != nil {
+			return err
+		}
+		for k := range tvs {
+			tvToDelete[k] = struct{}{}
+		}
+	}
+	go func() {
+		for {
+			for k := range templates {
+				if err := m.templateClient.Delete(k, &metav1.DeleteOptions{}); err != nil && !kerrors.IsNotFound(err) {
+					logrus.Warnf("Deleting template %v doesn't succeed. Continue loop", k)
+					continue
+				}
+			}
+			for k := range tvToDelete {
+				if err := m.templateVersionClient.Delete(k, &metav1.DeleteOptions{}); err != nil && !kerrors.IsNotFound(err) {
+					logrus.Warnf("Deleting templateVersion %v doesn't succeed. Continue loop", k)
+					continue
+				}
+			}
+			break
+		}
+	}()
+	return nil
+}
+
+func getCatalogType(cmt *CatalogInfo) string {
+	if cmt.projectCatalog == nil && cmt.clusterCatalog == nil {
+		return client.CatalogType
+	} else if cmt.projectCatalog != nil {
+		return client.ProjectCatalogType
+	} else {
+		return client.ClusterCatalogType
+	}
+}
+
+func (m *Manager) update(cmt *CatalogInfo, catalogType string, templateName string, condition bool, updateOnly bool) (*CatalogInfo, error) {
```

Comment:
can we think of more descriptive name for the method, like what exactly is it updating?